src/Utilities/HasUtilitiesTrait.php

```diff
@@ -21,4 +21,31 @@ function ksortMultiDimensional(array $input) {
             }
         }
     }
+
+    /**
+     * This function merge $array2 into $array1 and return the merged array.
+     *
+     * @param array $array1
+     * @param array $array2
+     *
+     * @return array
+     */
+    function arrayMergeMultiDimensional(array $array1, array $array2)
+    {
+        $merged = $array1;
+
+        foreach ($array2 as $key => &$value)
+        {
+            if (is_array($value) && isset($merged[$key]) && is_array($merged[$key])) {
+                $merged[$key] = $this->arrayMergeMultiDimensional($merged[$key], $value);
+            } else if (is_numeric($key)) {
+                if (!in_array($value, $merged))
+                    $merged[] = $value;
+            } else {
+                $merged[$key] = $value;
+            }
+        }
+
+        return $merged;
+    }
```

Comment:
@antonvas don't remember the specific cases, but when I checked with @amyboyd it was having funny behaviour.
In first refactoring we end up not using it, now I've tried to use it, but got again some strange behaviour in deeper levels of arrays.